BREAKING CHANGE: update token refresh handling (#175)
carforyou/carforyou-api-client-pkg: README.md

| `optionServiceUrl`           | URL to [Options Service](https://option-service.preprod.carforyou.ch/swagger-ui/index.html)                         |
| `analyticsServiceUrl`        | URL to [Analytics Service](https://carforyou-analytics-service.preprod.carforyou.ch/swagger-ui/index.html)          |
| `userNotificationServiceUrl` | URL to [User Notification Service](https://user-notification-service.preprod.carforyou.ch/swagger-ui/index.html)    |
| `debug`                      | Set to `true` to `console.log` requests and API responses.                                                          |

### Authenticated Requests

#### Access Token

In order to perform authenticated requests you need to set a valid access token to the api client in the consumers code `ApiClient.setAccessToken(JWT)` The provided token will be passed as `Authentication` header to the API.

#### Refreshing Access Tokens

> Authenticated requests will try to issue a refresh token when they revive a 401 response form the remote server. In order to handle the process of refreshing the token the consumer must pass a handler function to the API-Client. (The function may vary depending on the environment the API-Client is used)

Reviewer: ```suggestion
Authenticated requests will try to issue a refresh token when they receive a 401 response form the remote server. In order to handle the process of refreshing the token the consumer must pass a handler function to the API-Client. (The function may vary depending on the environment the API-Client is used)
```

```
ApiClient.setTokenRefreshHandler(async () => {
  const response = await fetch(process.env.TOKEN_REFRESH_ROUTE)
  return response.json()
})

> ```

Reviewer: I might see the answer later in the code, but how is the case handled when this function is missing, but the API client runs into a 401. A helpful message would be appreciated over throwing a cryptic message :)

Reviewer: will we have spot-on types for this functions return value to make sure we don't fuck refreshing up?

Author: 1. My intention was to throw a meaningful error in the function added in the prototype https://github.com/carforyou/carforyou-api-client-pkg/pull/175/files#diff-c50cd8cbf813f34536ae74335bcf5724R56 so if one misses to add it the error shall point them to whats off

2. The function is typed pretty strict https://github.com/carforyou/carforyou-api-client-pkg/pull/175/files#diff-1c1a34a94ec474ad2092388c3650708eR65 let me know if you would add something on that.

Reviewer: thanks for double checking. I forgot to delete this reminder-to-self-comment when I reached the function and answered the question to myself ;)

| `userNotificationServiceUrl` | URL to [User Notification Service](https://user-notification-service.preprod.carforyou.ch/swagger-ui/index.html)    |
| `debug`                      | Set to `true` to `console.log` requests and API responses.                                                          |

### Authenticated Requests

Requests that need authorization, need to pass a valid JWT to authenticate the user on the requested resource.

#### Implementing authenticated requests

To authenticate a request, simply pass the `isAuthorizedRequest` option to the request helper (fetch/post/put/delete-Data) and the api client will add the `Authorization` header. For server side requests, don't forget to pass the options down to the helper function so the consumer can pass the `accessToken` when invoking the request.

Note: If one forgets to add the `isAuthorizedRequest` option, the access token will not be set as a header. On the consumer side, if the access token is not passed, the api client will throw an error.

#### Performing authenticated requests

Pass the access token as a request option to the api call `dummyFetchCall(data, {accessToken: JWT})`

#### Refreshing Access Tokens

> The consumer is responsible to ensure a valid token is passed to the request. The api client will pass the provided token as an `Authorization` header to the api call.

Reviewer: if you have the energy, it might be worth documenting how to do that. can be separate, no need to rush it

| `optionServiceUrl`           | URL to [Options Service](https://option-service.preprod.carforyou.ch/swagger-ui/index.html)                         |
| `analyticsServiceUrl`        | URL to [Analytics Service](https://carforyou-analytics-service.preprod.carforyou.ch/swagger-ui/index.html)          |
| `userNotificationServiceUrl` | URL to [User Notification Service](https://user-notification-service.preprod.carforyou.ch/swagger-ui/index.html)    |
| `debug`                      | Set to `true` to `console.log` requests and API responses.                                                          |

### Authenticated Requests

> Authenticated requests need to be handled in a different way depending on weather the request is performed server- or client-side.

Reviewer: ```suggestion
Authenticated requests need to be handled in a different way depending on whether the request is performed server- or client-side.
```

> Client-Side, we can use the api client as a singelton to store the token. Server-side the access token needs to be passed to the request in order

Reviewer: ```suggestion
Client-Side, we can use the api client as a singleton to store the token. Server-side the access token needs to be passed to the request in order
```

I was involved in this decision, but reading this, I'm thinking - but why! Is it worth putting the mental overhead on the consumer to decide which one is the right approach for what context?
How much worse would it be for a server-side consumer if they'd have to pass the token with every call as well? The app can still decide to store the token in some singleton-ish way

> to not expose user credentials to the global module scope and ensure the request is performed with the correct

Reviewer: incomplete sentence?

#### Implementing authenticated requests

> To implement a authenticated request, simply pass the `isAuthorizedRequest` option to the request helper (fetch/post/put/delete-Data) and the api client will add the `Authorization` header. For server side requests, don't forget to pass the options down to the helper function so the consumer can pass the `accessToken` when invoking the request.

Reviewer: ```suggestion
To authenticate a request, simply pass the `isAuthorizedRequest` option to the request helper (fetch/post/put/delete-Data) and the api client will add the `Authorization` header. For server side requests, don't forget to pass the options down to the helper function so the consumer can pass the `accessToken` when invoking the request.
```

Reviewer: reading this, I'd love to get rid of all the "if this then that"'s (refers to the comment above). maybe we can simplify server/client side handling?

I'm also a bit confused by passing `isAuthorizedRequest`. wouldn't we just pass the token and that would then authenticate the request?

#### Access Token Handling

##### Client-Side requests

Set the access token along with it's expiration date to the api client using the `setToken` function provided by the api client instance.

##### Server-Side requests

> Pass the access token as a request option to the api call `dummyFetchCall(data, {accessToken: JWT})`

Reviewer: wouldn't that be the client-side version? the title of this paragraph says it's for server-side
